### PR TITLE
Allow multiple gate sets in a device spec.

### PR DESCRIPTION
**Pull request by dstrain115**
- make the gate set parameter an iterable so that a device can support multiple gate sets.

### CODE REVIEW

**Comment by dstrain115** [comment]
Ping, could I get a review on this?  It's been sitting around for 3 weeks.

File: cirq/google/known_devices.py [hunk 1]
```diff
@@ -57,7 +57,7 @@ def _parse_device(s: str) -> Tuple[List[GridQubit], Dict[str, Set[GridQubit]]]:
 
 def create_device_proto_from_diagram(
         ascii_grid: str,
-        gate_set: serializable_gate_set.SerializableGateSet = None,
+        gate_sets: Iterable[serializable_gate_set.SerializableGateSet] = None,
```
Comment (maffoo) [inline cirq/google/known_devices.py]: nit: Should be `Optional[Iterable[...]]` to allow for `None`.

File: cirq/google/serializable_device_test.py [hunk 2]
```diff
@@ -264,4 +265,52 @@ def test_mixing_types():
...
+
+
+def test_multiple_gatesets():
+    halfPiGateSet = cirq.google.serializable_gate_set.SerializableGateSet(
+        gate_set_name='half_pi_gateset',
+        serializers=cgc.SINGLE_QUBIT_HALF_PI_SERIALIZERS,
+        deserializers=cgc.SINGLE_QUBIT_HALF_PI_DESERIALIZERS,
+    )
+    allAnglesGateSet = cirq.google.serializable_gate_set.SerializableGateSet(
+        gate_set_name='half_pi_gateset',
```
Comment (maffoo) [inline cirq/google/serializable_device_test.py]: nit: use a different name here.

File: cirq/google/serializable_device_test.py [hunk 3]
```diff
@@ -264,4 +265,52 @@ def test_mixing_types():
...
+        serializers=cgc.SINGLE_QUBIT_SERIALIZERS,
+        deserializers=cgc.SINGLE_QUBIT_DESERIALIZERS,
+    )
+    durations_dict = {'xy_pi': 20_000, 'xy_half_pi': 20_000, 'xy': 20_000}
+    spec = cirq.google.known_devices.create_device_proto_from_diagram(
+        "aa\naa", [allAnglesGateSet, halfPiGateSet], durations_dict)
+    dev = cg.SerializableDevice.from_proto(
+        proto=spec, gate_sets=[allAnglesGateSet, halfPiGateSet])
+    dev.validate_operation(cirq.X(cirq.GridQubit(0, 0)))
+    dev.validate_operation(cirq.X(cirq.GridQubit(1, 0)))
```
Comment (maffoo) [inline cirq/google/serializable_device_test.py]: Should validate an operation with arbitrary angle as well.

File: cirq/google/serializable_device_test.py [hunk 4]
```diff
@@ -264,4 +265,52 @@ def test_mixing_types():
...
+        'xy_half_pi': 10_000,
+        'exp_w': 53_000,
+        'exp_11': 11_000,
+        'meas': 14_141
+    }
+    spec = cirq.google.known_devices.create_device_proto_from_diagram(
+        "aa\naa", [cirq.google.gate_sets.XMON, halfPiGateSet], durations_dict)
+    with pytest.raises(ValueError):
+        _ = cg.SerializableDevice.from_proto(
+            proto=spec, gate_sets=[_JUST_MEAS, halfPiGateSet])
```
Comment (maffoo) [inline cirq/google/serializable_device_test.py]: It's not exactly clear what is the conflict here. It looks like the gate_sets will be missing the XMON gates, but that would seem to be a different error condition from conflicting definitions as implied by the test name.

**Comment by maffoo** [comment]
Sorry for the delayed review; looks good, just a few minor comments.

File: cirq/google/serializable_device.py [hunk 5]
```diff
@@ -40,6 +40,10 @@ def __init__(self, duration: Duration,
             q for qubit_tuple in target_set for q in qubit_tuple
         }
 
+    def __eq__(self, other):
+        return (isinstance(other, self.__class__) and
```
[the remark this answers is not in the document]
Comment (dstrain115) [reply cirq/google/serializable_device.py]: Done

File: cirq/google/known_devices.py (hunk 1 above)
Comment (dstrain115) [reply cirq/google/known_devices.py]: Done.

File: cirq/google/serializable_device_test.py (hunk 2 above)
Comment (dstrain115) [reply cirq/google/serializable_device_test.py]: Done

File: cirq/google/serializable_device_test.py (hunk 3 above)
Comment (dstrain115) [reply cirq/google/serializable_device_test.py]: Done

File: cirq/google/serializable_device_test.py (hunk 4 above)
Comment (dstrain115) [reply cirq/google/serializable_device_test.py]: Changed this test around to make it more readable and more robust.

Also, it was depending on XMON and SINGLE_QUBIT to serialize into different symbols, hence the conflict, but I am removing that distinction in a different PR.

File: cirq/google/devices/serializable_device.py [hunk 6]
```diff
@@ -139,6 +141,9 @@ def from_proto(
                         raise ValueError(f'Serializer has {gate_id} which is '
                                          'not supported by the device '
                                          'specification')
+                    if (gate_type in gates_by_type):
+                        print(gate_definitions[gate_id].__dict__)
+                        print(gates_by_type[gate_type].__dict__)
```
Comment (maffoo) [inline cirq/google/devices/serializable_device.py]: Did you mean to leave this in?

Comment (dstrain115) [reply cirq/google/devices/serializable_device.py]: Nope.  Removed.

**Pull request merged**